Password Security: reviewed version (#411)
hyperledger/iroha-2-docs: src/guide/security/password-security.md

# Password Security

In the realm of blockchain security, protecting passwords is paramount. To ensure your data and everything it represents remain impervious to unauthorized access, let's delve into the nuances of password security.

## Password Strength

Likely enough, you may have previously encountered recommendations on how to come up with a _strong_ password. These may entail such advice as minimum password length, addition of special characters, etc. Such recommendations aim to increase the strength of your password that hinges on entropy, i.e. randomness of the password.

So, what defines a _strong password_? A strong password is a password with _high entropy_.

To calculate the entropy of a password, we may follow the **Entropy formula**:

::: tip Entropy formula

$L$ — Password length; number of symbols in the password.\
$S$ — Character set; size of the pool of unique possible symbols.\
$S^L$ — Number of possible combinations.

\[ Entropy=log_2(S^L) \]

The resulting number is the amount of entropy bits in a password. The higher the number, the harder the password is to crack.

> Having an entropy value, the amount of attempts required to brute-force a password with said entropy can be derived by using the following formula:

Reviewer: ```suggestion
The amount of attempts required to brute-force a password with given entropy can be derived by using the following formula:
```

\[ S^L=2^Entropy \]

> There is no universal answer as to how high the entropy of a password should be. For financial organizations, it is advised to keep the entropy of their passwors in the range from `64` to `127` bits. However, keep in mind that GPUs keep constantly evolving, and the time required for password cracking keeps decreasing over time.

Reviewer: ```suggestion
There is no universal answer as to how high the entropy of a password should be. For financial organizations, it is advised to keep the entropy of their passwords in the range from `64` to `127` bits. However, keep in mind that GPUs keep constantly evolving, and the time required for password cracking keeps decreasing over time.
```

:::

Following the entropy formula, let us compare the following two examples:

  1. A 16-character password with the character set utilizing only lowercase letters of the modern English alphabet (26 characters) yields approximately 43 sextillion ($43*10^21$) possible combinations.

    \[ Entropy=log_2(26^{16})=log_2(43,608,742,899,428,874,059,776)=75.20703... \]

  2. A 16-character password with the character set expanded to 96, including uppercase letters and special symbols, inflates the number of possible combinations to a staggering 52 nonillion ($52*10^30$), improving entropy significantly.

    \[ Entropy=log_2(96^{16})=log_2(52,040,292,466,647,269,602,037,015,248,896)=105.35940... \]

As can be seen, even by only expanding the character set from 26 to 96 symbols, the number of possible combinations that a malicious party would need to bruteforce has expanded by $1.1933*10^9$ times.

> Additionally increasing the length of the password, will grow the number of possible combinations even further, therefore enhancing the entropy—strength—of the password.

Reviewer 2: ```suggestion
Additionally, increasing the password length will grow the number of possible combinations even further, enhancing the entropy—strength—of the password.
```

> However, instead of wrestling with complexities, we advise using a password manager program—like [KeePassXC](https://keepassxc.org/) (for more details, see _[Adding a Password Manager Program](./storing-cryptographic-keys.md#adding-a-password-manager-program)_ and _[Configuring KeePassXC](./storing-cryptographic-keys.md#configuring-keepasxc)_)—to generate and securely store your passwords.

Reviewer 2: ```suggestion
However, instead of wrestling with complexities, we advise using a password manager program, like [KeePassXC](https://keepassxc.org/) (for more details, see _[Adding a Password Manager Program](./storing-cryptographic-keys.md#adding-a-password-manager-program)_ and _[Configuring KeePassXC](./storing-cryptographic-keys.md#configuring-keepasxc)_)—to generate and securely store your passwords.
```
